--- base\forge\fml\mc_versions.cfg	Sun Jul  7 15:14:54 2013
+++ work\forge\fml\mc_versions.cfg	Fri May 30 21:39:39 2014
@@ -109,7 +109,7 @@
 mcp_md5    = 7810c2fba33e15889685b21ac8dc66cf
 
 [1.6.2]
-client_md5 = 9356f3ad44925a684e918b9270b63a6c
+client_md5 = 1d43cdba8b9752d7f557ab7d3e365964
 server_md5 = 39df9f29e6904ea7b351ffb4fe949881
 mcp_ver    = 8.04
 mcp_url    = http://mcp.ocean-labs.de/files/archive/mcp804.zip
