--- base\forge\fml\mc_versions.cfg	Wed Nov 13 19:21:34 2013
+++ work\forge\fml\mc_versions.cfg	Tue Apr 29 16:08:49 2014
@@ -124,7 +124,7 @@
 
 [1.6.4]
 client_md5 = 2e5044f5359e82245551167a237f3167
-server_md5 = ba3145656b1480122bd8759cecd7b7a1
+server_md5 = abcf286a14f7aee82e8bf89270433509
 mcp_ver    = 8.11
 mcp_url    = http://mcp.ocean-labs.de/files/archive/mcp811.zip
 mcp_md5    = 2d7a759309b5cc10ca29caa0b10f3bfc
